docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,154 @@
+  Users can implement this function to start a tracking session at the earliest possible
+  time.
+
+Please note that the timing of calling of these functions is important. For
+example, the starting and stoping tracking must cover the coupled malloc/free
+operations, which means start tracking before malloc and also stop tracking
+after free, or debug malloc will definitely trigger false alarms. And by the same
+logic, oe_debug_malloc_print_objects() can only be called after tracking is
+stopped.
+
+# Implementation
+
+## Building OE SDK from source retains current behavior.
+
+OE SDK build system will retain the USE_DEBUG_MALLOC cmake parameter.
+By default, in debug build, it is set to `on` and enclaves will link against
+oe_debug_malloc and enable leak detection.
+
+The default implementation of oe_debug_malloc_init_callback will be thus:
+```c
+void oe_debug_malloc_init_callback(void)
+{
+#ifdef OE_USE_DEBUG_MALLOC // defined when USE_DEBUG_MALOC=on
+   oe_debug_malloc_start_tracking();
+#endif
+}
+```
+Of course, this default setting can be turned off by changing USE_DEBUG_MALLOC
+to `off`.
+
+## User-defined callback function to subsitute the predefined function
+While user defines the callback function following the definition and matches
+the name exactly, the predefined function will be substituted by user's, thus
+user can decide the point to start tracking by calling the callback function.
+
+The expected implementation of the callback function is as following:
+  ```c
+  void oe_debug_malloc_init_callback(void)
+  {
+      oe_debug_malloc_start_tracking();
+  }
+  ```
+
+# Sample
+
+```c
+#include <openenclave/debugmalloc.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <stdio.h>
+
+// Include the trusted debugmalloc header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the debugmalloc.edl file.
+#include "debugmalloc_t.h"
+
+// Callback function to initiate debugmalloc
+void oe_debug_malloc_init_callback(void)
+{
+    oe_debug_malloc_start_tracking();
+}
+
+// This is the function that the host calls. It prints
+// a message in the enclave before calling back out to
+// the host to print a message from there too.
+void enclave_hello()
+{
+    oe_debug_malloc_init_callback();
+
+    void* buf = oe_malloc(4096);
+    oe_free(buf);
+
+    oe_debug_malloc_stop_tracking();
+
+    oe_debug_malloc_print_objects();
+
+    // Call back into the host
+    oe_result_t result = host_hello();
+    if (result != OE_OK)
+    {
+        fprintf(
+            stderr,
+            "Call to host_hello failed: result=%u (%s)\n",
+            result,
+            oe_result_str(result));
+    }
+}
+```
+While this sample has no memory leak, there is no expected extra output.
+However if there is memory leaking, oe_debug_malloc_print_objects() can report
+the leaks, and the enclave will raise an error "OE_MEMORY_LEAK".
+
+# Note
```

Comment:
It would be useful to document how this new feature interacts with the pluggable allocators:

- It is fully composable with a pluggable OE allocator
- It does not require an OE allocator to implement new or different APIs to be used with the debug malloc.
- It does not require or guarantee any thread synchronization/serialization behaviors when used with thread-aware allocators.

Reply:
Thanks, partially updated.